pkg/apis/core/validation/validation.go

```diff
@@ -6974,3 +7047,264 @@ func sameLoadBalancerClass(oldService, service *core.Service) bool {
+	}
+	allErrs = append(allErrs, validateResourceClaimParameters(spec.ParametersRef, fldPath.Child("parametersRef"))...)
+	switch spec.AllocationMode {
+	case core.AllocationModeImmediate, core.AllocationModeWaitForFirstConsumer:
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("allocationMode"), spec.AllocationMode, "unknown mode"))
+	}
+	return allErrs
+}
+
+// It would have been nice to use Go generics to reuse the same validation
+// function for Kind and Name in both types, but generics cannot be used to
+// access common fields in structs.
+
+func validateResourceClaimParameters(ref *core.ResourceClaimParametersReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if ref != nil {
+		if ref.Kind == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("kind"), "must be specified"))
+		}
+		if ref.Name == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("name"), "must be specified"))
+		}
+	}
+	return allErrs
+}
+
+func validateResourceClassParameters(ref *core.ResourceClassParametersReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if ref != nil {
+		if ref.Kind == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("kind"), "must be specified"))
+		}
+		if ref.Name == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("name"), "must be specified"))
+		}
+		if ref.Namespace != "" {
+			for _, msg := range ValidateNamespaceName(ref.Namespace, false) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), ref.Namespace, msg))
+			}
+		}
+	}
+	return allErrs
+}
+
+// ValidateResourceClass validates a ResourceClass.
+func ValidateResourceClass(resourceClass *core.ResourceClass) field.ErrorList {
+	allErrs := ValidateObjectMeta(&resourceClass.ObjectMeta, false, ValidateClassName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateResourceDriverName(resourceClass.DriverName, field.NewPath("driverName"))...)
+	allErrs = append(allErrs, validateResourceClassParameters(resourceClass.ParametersRef, field.NewPath("parametersRef"))...)
+
+	return allErrs
+}
+
+// ValidateResourceClassUpdate tests if an update to ResourceClass is valid.
+func ValidateResourceClassUpdate(resourceClass, oldResourceClass *core.ResourceClass) field.ErrorList {
+	allErrs := ValidateObjectMetaUpdate(&resourceClass.ObjectMeta, &oldResourceClass.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateResourceClass(resourceClass)...)
+	return allErrs
+}
+
+// ValidateResourceClaimUpdate tests if an update to ResourceClaim is valid.
+func ValidateResourceClaimUpdate(resourceClaim, oldResourceClaim *core.ResourceClaim) field.ErrorList {
+	allErrs := ValidateObjectMetaUpdate(&resourceClaim.ObjectMeta, &oldResourceClaim.ObjectMeta, field.NewPath("metadata"))
+	if !apiequality.Semantic.DeepEqual(oldResourceClaim.Spec, resourceClaim.Spec) {
```

Comment:
Perhaps the Romans were right in this case: ValidateImmutableField reports the entire struct as immutable, without pointing out exactly which field in it was modified.

```
            	            	(field.ErrorList) (len=1) {
            	            	 (*field.Error)({
            	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
            	            	  Field: (string) (len=4) "spec",
            	            	  BadValue: (core.ResourceClaimSpec) {
            	            	   ResourceClassName: (string) (len=5) "valid",
            	            	   ParametersRef: (*core.ResourceClaimParametersReference)({
            	            	    APIGroup: (string) "",
            	            	    Kind: (string) (len=3) "foo",
            	            	    Name: (string) (len=3) "bar"
            	            	   }),
            	            	   AllocationMode: (core.AllocationMode) (len=20) "WaitForFirstConsumer"
            	            	  },
            	            	  Detail: (string) (len=18) "field is immutable"
            	            	 })
```

The solution with `cmp` is more user-friendly. I am going to stick with ValidateImmutableField because it is designed for this purpose. Perhaps it can be enhanced at some later point in time.

Reply:
I think we should change the function to use cmp.Diff indeed (but follow up PR)

Reply:
https://github.com/kubernetes/kubernetes/issues/113633